### PR TITLE
feat(gateway): watch Gateway for tag changes in GatewayInstance reconciler

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

Closes https://github.com/kumahq/kuma/issues/3509

### CODE REVIEW

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 1]
```diff
@@ -17,9 +17,13 @@ import (
 	kube_runtime "k8s.io/apimachinery/pkg/runtime"
 	kube_schema "k8s.io/apimachinery/pkg/runtime/schema"
 	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_record "k8s.io/client-go/tools/record"
 	kube_ctrl "sigs.k8s.io/controller-runtime"
 	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
 	kube_controllerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	kube_handler "sigs.k8s.io/controller-runtime/pkg/handler"
+	kube_reconile "sigs.k8s.io/controller-runtime/pkg/reconcile"
```
Comment (jpeach) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:25]:
```suggestion
	kube_reconcile "sigs.k8s.io/controller-runtime/pkg/reconcile"
```

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 2]
```diff
@@ -31,9 +35,16 @@ import (
 	util_k8s "github.com/kumahq/kuma/pkg/plugins/runtime/k8s/util"
 )
 
+const (
+	// FailedToCreateService is added to an event when
+	// the Service for a GatewayInstance can't be created.
+	FailedToCreateService = "FailedToCreateService"
```
Comment (jpeach) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:41]: I'm not sure this is the right event name to publish. As used, it doesn't indicate there was a failure (we didn't attempt to create a service and have that attempt fail), it indicates that we should not be creating a service and so we didn't, which is OK and normal.

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 3]
```diff
@@ -308,6 +333,43 @@ func updateStatus(gatewayInstance *mesh_k8s.GatewayInstance, svc *kube_core.Serv
...
+			context.Background(), instances, kube_client.MatchingFields{serviceKey: serviceName},
+		); err != nil {
+			l.WithValues("gateway", obj.GetName()).Error(err, "failed to fetch GatewayInstances")
+			return nil
+		}
+
+		var req []kube_reconile.Request
+		for _, instance := range instances.Items {
+			req = append(req, kube_reconile.Request{
+				NamespacedName: kube_types.NamespacedName{Namespace: instance.Namespace, Name: instance.Name},
```
Comment (jpeach) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:365]: What would happen when a Gateway changes its service tags such that it unbinds from a GatewayInstance?

**Review by jpeach — CHANGES_REQUESTED** [review]
I think that the commit and PR message buries the lede a bit. AFAICT the main change here is to watch gateways so that we reconcile the GatewayInstance when a binding could be made. Publishing whether there is a binding is a secondary benefit.

**Comment by michaelbeaumont** [comment]
> I think that the commit and PR message buries the lede a bit. AFAICT the main change here is to watch gateways so that we reconcile the GatewayInstance when a binding could be made. Publishing whether there is a binding is a secondary benefit.

Definitely, that's because the requeuing change came after I realized it'd never be requeued. I'll update.

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 4]
```diff
@@ -308,6 +333,43 @@ func updateStatus(gatewayInstance *mesh_k8s.GatewayInstance, svc *kube_core.Serv
...
 
+const serviceKey string = ".metadata.service"
+
+func GatewayToInstanceMapper(l logr.Logger, client kube_client.Client) kube_handler.MapFunc {
+	l = l.WithName("gateway-to-gateway-instance-mapper")
+
+	return func(obj kube_client.Object) []kube_reconile.Request {
+		gateway := obj.(*mesh_k8s.Gateway)
+
+		// Take first service from selectors
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:344]: We want to match the index of GatewayInstances, which holds the the value of the server tag.

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go (hunk 3 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:293]: [`controller-runtime enqueues both the old and the new object for an Update`](https://github.com/kubernetes-sigs/controller-runtime/blob/f236f0345ad2933912ebf34bfcf0f93620769654/pkg/handler/enqueue_mapped.go#L63-L64) so the unbound-from `GatewayInstance` will be enqueued.

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jpeach) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:271]: OK, but a Gateway can have a different service tag in each selector, and each of those can match to a GatewayInstance. In that case, when the Gateway changes, we would only update one of the matching GatewayInstance resources, right?

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go (hunk 3 above)
Comment (jpeach) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:291]: OK, so `GatewayToInstanceMapper` will get called twice with the old and new objects. When it gets called with the old, then we should find the old binding and enqueue and update for that, and when it gets called with the new we would find the new binding. The final state would then be updated my the main reconciler.

If that's correct, I think it's worth leaving a comment.

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:271]: True. For some reason I had it in my head that only one service can be selected by a Gateway but of course the requirement is that each service can be selected by only one Gateway.

**Review by jpeach — APPROVED** [review]
Nice one. LGTM.

**Pull request merged**